### PR TITLE
[MO][OPS-11742] Integrate NPS Claim Overpayment API

### PR DESCRIPTION
**Pull request by Lyrain**

### CODE REVIEW

File: test/pagespecs/VerifyBankAccountPageSpec.scala [hunk 1]
```diff
@@ -59,6 +59,12 @@ class VerifyBankAccountPageSpec extends ItSpec {
     pages.verifyBankAccountPage.assertPageIsDisplayed()
```
Comment (barnesjake) [inline test/pagespecs/VerifyBankAccountPageSpec.scala:60]: I think we should add a test scenario for when the claim overpayment api returns something other than 200, just so we have behaviour documented in tests

File: app/nps/ClaimOverpaymentConnector.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ClaimOverpaymentConnector @Inject() (
+    npsConfig:  NpsConfig,
+    httpClient: HttpClient
+)(implicit ec: ExecutionContext) {
+
+  private def url(nino: Nino, p800Reference: P800Reference): String = npsConfig.baseUrl +
+    s"/nps-json-service/nps/v1/api/accounting/claim-overpayment/${nino.value}/${p800Reference.value}"
```
Comment (pawelpanasewicz) [inline app/nps/ClaimOverpaymentConnector.scala:37]: p800Reference.sanitiseReference.value (otherwise it might contain non digits)

Comment (barnesjake) [reply app/nps/ClaimOverpaymentConnector.scala:37]: Very good spot Pawel! I completely missed this

File: app/controllers/WeAreVerifyingYourBankAccountController.scala [hunk 3]
```diff
@@ -63,19 +67,57 @@ class WeAreVerifyingYourBankAccountController @Inject() (
...
+          case otherResult =>
+            Errors.throwServerErrorException(s"Unexpected claim overpayment result. This should be investigated. [otherResult: ${otherResult.toString}]")
+        }
+      }
       case EventValue.NotValid    => Redirect(routes.RefundRequestNotSubmittedController.get)
       case EventValue.NotReceived => Ok(views.weAreVerifyingYourBankAccountPage(status, consent_id, bank_reference_id))
     }
   }
 
+  private def claimOverpayment(journey: Journey, isValid: EventValue)(implicit request: JourneyRequest[_]): Future[_] = isValid match {
```
Comment (pawelpanasewicz) [inline app/controllers/WeAreVerifyingYourBankAccountController.scala:96]: The function should not take `isValid` into account. The check is already done in outer scope. 
(I can change it when merging my changes)

File: app/nps/ClaimOverpaymentConnector.scala [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    val sanitisedP800Reference = p800Reference.sanitiseReference
+    implicit val reads: HttpReads[ClaimOverpaymentResult] = ClaimOverpaymentConnector.reads
+
+    httpClient
+      .PUT[ClaimOverpaymentRequest, ClaimOverpaymentResult](
+        url     = url(nino, sanitisedP800Reference),
+        body    = claimOverpaymentRequest,
+        headers = npsConfig.makeHeadersForNps()
+      ).map {
+          case claimOverpaymentResult: ClaimOverpaymentResult => claimOverpaymentResult
```
Comment (pawelpanasewicz) [inline app/nps/ClaimOverpaymentConnector.scala:60]: I'm not sure what it does. Is it needed.

File: app/nps/ClaimOverpaymentConnector.scala [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+   * 4. Any other error code status.
+   *
+   * Cases 2, 3 should be handled gracefully.
+   * Case 4 results in 'Technical difficulties' page.
+   */
+  val reads: HttpReads[ClaimOverpaymentResult] = HttpReads.ask.flatMap {
+    case (method, url, response) => response.status match {
+      case Status.OK => HttpReads[ClaimOverpaymentResult.ClaimOverpaymentResponse].map((x: ClaimOverpaymentResult.ClaimOverpaymentResponse) => x: ClaimOverpaymentResult)
+      case Status.UNPROCESSABLE_ENTITY =>
+        logger.info(s"UNPROCESSABLE_ENTITY: ${response.body}")
```
Comment (pawelpanasewicz) [inline app/nps/ClaimOverpaymentConnector.scala:82]: Please remove this logger statement from reads.

File: app/nps/models/ClaimOverpaymentResponse.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+object ClaimOverpaymentResult {
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  implicit val format: OFormat[ClaimOverpaymentResult] = derived.oformat[ClaimOverpaymentResult]()
+
+  case object RefundAlreadyTaken extends ClaimOverpaymentResult
+
+  case object RefundSuspended extends ClaimOverpaymentResult
+
```
Comment (pawelpanasewicz) [inline app/nps/models/ClaimOverpaymentResponse.scala:33]: RefundAlreadyTaken and RefundSuspended aren't needed IMO. The ClaimOverpaymentResult should be only a case class. If the endpoint results in 422 we simply treat it as unexpected response and throw exception and render technical difficulties.

File: app/nps/ClaimOverpaymentConnector.scala [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+   *
+   * 1. 200 status code with happy path JSON ClaimOverpaymentResponse
+   * 2. 422 status code, with a JSON failure code mapping to 'Overpayment already taken'
+   * 3. 422 status code, with a JSON failure code mapping to 'Overpayment suspended'
+   * 4. Any other error code status.
+   *
+   * Cases 2, 3 should be handled gracefully.
+   * Case 4 results in 'Technical difficulties' page.
+   */
+  val reads: HttpReads[ClaimOverpaymentResult] = HttpReads.ask.flatMap {
```
Comment (pawelpanasewicz) [inline app/nps/ClaimOverpaymentConnector.scala:78]: I think the custom reads aren't needed.

RefundAlreadyTaken and RefundSuspended aren't needed IMO. If we remove them this reads becomes auto generated from Json macro.

**Review by pawelpanasewicz — APPROVED** [review]

**Pull request merged**